lib/prometheus/engineering_collector.rb

```diff
@@ -28,13 +28,16 @@ def type
   end
 
   def dangling(model)
-    model.where.not(host_id: Host.select(:id))
+    model.where.not(host_id: Host.with_policy_or_test_results.select(:id))
+  end
+
+  def dangling_accounts
+    Account.where.not(account_number: Host.with_policy_or_test_results
+                                          .select(:account).distinct)
   end
 
   def collect
```

Comment:
Would it be possible to reduce the interval of collection of these expensive queries, or an introduction of some caching?

Reply:
The real silver bullet here would be to add a read only instance of the DB for queries like these that cannot affect performance of prod for our users. The /metrics endpoint is polled; I'm not sure if we can configure the polling interval. If you see some opportunity for caching, please point it out, though the data collected here are changing constantly (kind of the point of metrics).

Reply:
I guess you've meant to have a read-only copy of the database synced with the main one. That's also an option. For our purposes it might be an overkill.

For caching we can leverage Redis to cache it for some time period. We don't need to have immediate value available at all times. We would not miss out any spikes that might happen during the cache period, would we?

Reply:
>I guess you've meant to have a read-only copy of the database synced with the main one. That's also an option. For our purposes it might be an overkill.

Yes, exactly. I believe it is supported directly from app-interface's managed resources (i.e. an out of the box solution for RDS replication). I agree it's overkill at this point considering the amount of traffic we have. With the appropriate index in place on accounts, even joining with 600k hosts should not bring RDS to its knees.

I have been thinking about this 50 second timeout, though. It would seem to make sense to decrease it to something reasonable like 5 seconds. If a query result doesn't return in 5 seconds, there is something seriously wrong with the state of our DB or with the query itself. In any case, polling seems to happen (from prometheus /metrics logs) every 5 seconds, so a timeout greater than that makes little sense:

```
10.130.2.1 - - [24/Mar/2021:15:30:05 UTC] "GET /metrics HTTP/1.1" 200 32640
- -> /metrics
10.130.20.31 - - [24/Mar/2021:15:30:09 UTC] "GET /metrics HTTP/1.1" 200 4435
- -> /metrics
10.130.2.1 - - [24/Mar/2021:15:30:14 UTC] "GET /metrics HTTP/1.1" 200 32640
- -> /metrics
10.129.4.104 - - [24/Mar/2021:15:30:19 UTC] "GET /metrics HTTP/1.1" 200 4433
- -> /metrics
```

Reply:
btw, I'm curious if you think a smaller timeout would have saved the DB from being overloaded. If the caller times out, I guess it would depend on if some message is sent to the DB to cancel the pending query or not. I'm not sure how the timeout works.

Reply:
I'm not sure if the timeout on the exporter would help us much with what has happened. I've also believed that the queries are being terminated if the caller closes the connection, but from what we've seen the queries kept hanging in RDS. I could have been due to several factors like the exporter pods being terminated and/or the use of `dblink`.

Anyhow, a lower timout would definitely help with some cases. At least if there would be several long running queries, it could stop at the first one.

Does the exporter run on a single worker/thread, because if so, the longer timeout on collection could hold/timeout requests coming in the interim.

Reply:
>Does the exporter run on a single worker/thread, because if so, the longer timeout on collection could hold/timeout requests coming in the interim.

Yes, it appears it's single threaded from my tests locally. I also don't see any option for setting more threads/workers.

>I've also believed that the queries are being terminated if the caller closes the connection, but from what we've seen the queries kept hanging in RDS.

There may be different behavior on timeout vs. if the pod is killed. I'd expect the timeout to clean up more, but the pod getting killed almost certainly cannot clean up any open connections. Also, the timeout should still use the same connection to the DB rather than creating a new pod with a new connection and leaving the old connection dangling. There is a configurable maximum number of DB connections available (50 in prod, I think?).